### PR TITLE
prom-query-fixes

### PR DESCRIPTION
**Pull request by yaronha**

### CODE REVIEW

File: pkg/querier/querier.go [hunk 1]
```diff
@@ -36,25 +36,39 @@ import (
...
 func NewV3ioQuerier(container *v3io.Container, logger logger.Logger, mint, maxt int64,
 	cfg *config.V3ioConfig, partMngr *partmgr.PartitionManager) *V3ioQuerier {
-	newQuerier := V3ioQuerier{container: container, mint: mint, maxt: maxt,
-		logger: logger.GetChild("Querier"), cfg: cfg}
+	newQuerier := V3ioQuerier{
+		container: container,
+		mint:      mint, maxt: maxt,
+		logger:            logger.GetChild("Querier"),
+		cfg:               cfg,
+		disableClientAggr: cfg.DisableClientAggr,
```
Comment (imakhlin) [inline pkg/querier/querier.go]: Why did you add this attribute? `cfg` already has it.

File: pkg/querier/querier.go [hunk 2]
```diff
@@ -36,25 +36,39 @@ import (
...
+	disableAllAggr    bool
 }
 
-// Standard Time Series Query, return a set of series which match the condition
+// Prometheus Time Series Query, return a set of series which match the condition
 func (q *V3ioQuerier) Select(name, functions string, step int64, filter string) (SeriesSet, error) {
 	return q.selectQry(name, functions, step, nil, filter)
 }
 
+// Standard Time Series Query, return a set of series which match the condition
```
Comment (imakhlin) [inline pkg/querier/querier.go]: `Prometheus`

File: pkg/querier/querier.go [hunk 3]
```diff
@@ -36,25 +36,39 @@ import (
...
+	container         *v3io.Container
+	cfg               *config.V3ioConfig
+	mint, maxt        int64
+	partitionMngr     *partmgr.PartitionManager
+	disableClientAggr bool
+	disableAllAggr    bool
 }
 
-// Standard Time Series Query, return a set of series which match the condition
+// Prometheus Time Series Query, return a set of series which match the condition
```
Comment (imakhlin) [inline pkg/querier/querier.go]: `Standard`

File: pkg/querier/series.go [hunk 4]
```diff
@@ -106,6 +106,8 @@ func (s *V3ioSeries) initSeriesIter() {
 				s.set.logger.ErrorWith("Error reading chunk buffer", "Lset", s.lset, "err", err)
 			} else {
 				newIterator.chunks = append(newIterator.chunks, chunk)
+				newIterator.chunksMax = append(newIterator.chunksMax,
```
Comment (imakhlin) [inline pkg/querier/series.go]: if both arrays are of the same size, consider using a struct and do append at once

File: pkg/aggregate/iterator.go [hunk 5]
```diff
@@ -71,7 +71,7 @@ func (as *AggregateSeries) CanAggregate(partitionAggr AggrType) bool {
 	// make sure the DB has all the aggregators we need (on bits in the mask)
 	// and that the requested interval is greater/eq to aggregator resolution and is an even divisor
 	return ((aggrMask & partitionAggr) == aggrMask) &&
-		as.interval >= as.rollupTime && (as.interval%as.rollupTime == 0)
+		as.interval >= as.rollupTime && (as.interval%as.rollupTime == 0 || as.interval/as.rollupTime > 3)
```
Comment (talIguaz) [inline pkg/aggregate/iterator.go]: What exactly are you checking here?

File: pkg/querier/seriesset.go [hunk 6]
```diff
@@ -100,7 +102,7 @@ func (s *V3ioSeriesSet) Next() bool {
 
 		s.nullSeries = false
 
-		if s.aggrSeries.CanAggregate(s.partition.AggrType()) && s.maxt-s.mint >= s.interval {
+		if s.canAggregate {
```
Comment (talIguaz) [inline pkg/querier/seriesset.go]: you sure you want to delete this condition ` && s.maxt-s.mint >= s.interval`? why isn't it relevant anymore?

File: promtsdb/promtsdb.go [hunk 7]
```diff
@@ -56,10 +56,21 @@ type V3ioPromQuerier struct {
...
 	name, filter, functions := match2filter(oms)
+	noAggr := false
+
 	if params.Func != "" {
-		functions = params.Func
+		// only pass xx_over_time functions (just the xx part)
+		// TODO: support count/stdxx, require changes in Prometheus: promql/functions.go, not calc aggregate twice
+		f := params.Func
+		if strings.HasSuffix(f, "_over_time") {
+			if params.Step == 0 && (f[0:3] == "min" || f[0:3] == "max" || f[0:3] == "sum" || f[0:3] == "avg") {
```
Comment (talIguaz) [inline promtsdb/promtsdb.go]: better initialize a variable with `f[0:3]` and use it. it's more readable and understandable.

File: pkg/querier/querier.go (hunk 1 above)
Comment (yaronha) [reply pkg/querier/querier.go]: @imakhlin this is the default value, may be overwritten later (e.g. by promtsdb)

File: pkg/aggregate/iterator.go (hunk 5 above)
Comment (yaronha) [reply pkg/aggregate/iterator.go]: @imakhlin explained to Tal, its handling prom step rounding errors issues

File: pkg/querier/seriesset.go (hunk 6 above)
Comment (yaronha) [reply pkg/querier/seriesset.go]: @imakhlin not relevant (old condition)

File: promtsdb/promtsdb.go (hunk 7 above)
Comment (yaronha) [reply promtsdb/promtsdb.go]: changed

File: pkg/aggregate/iterator.go (hunk 5 above)
Comment (talIguaz) [reply pkg/aggregate/iterator.go]: Better add a comment that explains that condition, this condition isn't really self explanatory
